api/network_cnm_compat.go

```diff
@@ -0,0 +1,99 @@
+package api
+
+// Convenience functions for dealing with NetworkSpec and Network
+// objects with compat CNM fields.
+
+import (
+	"github.com/docker/swarmkit/api/deepcopy"
+)
+
+// GetCNMCompat is like GetCNM but if necessary falls back to legacy
+// fields in the NetworkSpec. Returns `nil` if the Network object is
+// not a CNM object.
+//
+// Callers should not expect modifications to the returned structure
+// to update the NetworkSpec and it is not expected that anything will
+// ever need to write those fields after they are set by the
+// originating client. Hence no corresponding SetCNMCompat method is
+// provided.
+func (ns *NetworkSpec) GetCNMCompat() *CNMNetworkSpec {
+	switch t := ns.Backend.(type) {
+	case *NetworkSpec_CNM:
+		// This is a bit more expensive than simply returning
+		// t.CNM but avoids callers accidentally relying on
+		// being able to write this struct.
+		var spec CNMNetworkSpec
+		deepcopy.Copy(&spec, t.CNM)
+		return &spec
+	case nil:
+		return &CNMNetworkSpec{
+			DriverConfig: ns.CNMCompatDriverConfig,
+			Ipv6Enabled:  ns.CNMCompatIpv6Enabled,
+			Internal:     ns.CNMCompatInternal,
+			IPAM:         ns.CNMCompatIPAM,
+			Attachable:   ns.CNMCompatAttachable,
+			Ingress:      ns.CNMCompatIngress,
+		}
+	default:
+		return nil
+	}
+}
+
+// GetCNMCompat is like GetCNM but if necessary falls back to legacy
+// fields in the Network object. Returns `nil` if the Network object
+// is not a CNM object.
+//
+// Callers which modify the returned state object and wish for those
+// modifications to persist _must_ call SetCNMCompat() after making
+// their changes. (perhaps by calling "defer n.SetCNMCompat(state)"?)
+//
+// Callers should not hold the returned CNMState over calls to the
+// allocator if they care about freshness, they should call
+// GetCNMCompat again to obtain a fresh copy.
+func (n *Network) GetCNMCompat() *CNMState {
+	// The location of the state (compat or not) always
+	// corresponds to the compat status of the spec. Hence we
+	// check n.Spec not n.State
+	switch n.Spec.Backend.(type) {
+	case *NetworkSpec_CNM:
+		if n.State == nil { // State not yet set?
+			return &CNMState{}
+		}
+		state, ok := n.State.(*Network_CNM)
+		if !ok {
+			panic("Unexpected non CNM state on CNM Network")
```

Comment:
I made this change and did a pass over all callers and ensured that they propagated the nil into an error return.